### ISSUE TITLE
`cursor_bookmark` is no longer working as expected for inserting text in tables

### ISSUE THREAD

**Issue by aghaynes** [davidgohel/officer#539]
I believe that there is a bug with the bookmarks, particularly when they are placed in tables. In version 0.4.3, the text replaced the bookmark in the table. In 0.6.3 the text is placed after the table. Bookmarks outside of tables seem to work as expected.

- [x] Provide the code that is producing the error, it has to be a minimal reproducible example.
Stackoverflow is providing good explanations about it: https://stackoverflow.com/help/mcve. You can use package `reprex` to help you: http://reprex.tidyverse.org/. The most popular R stackoverflow question is about the subject: https://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example.

Here is a small template docx file with a few bookmarks: 
[bm.docx](https://github.com/davidgohel/officer/files/13362541/bm.docx)

The following code should replace the bookmarks:
```
library(officer)

doc <- read_docx("bm.docx")

doc <- cursor_bookmark(doc, "a1") |> 
  body_add_par("Text for A1") |> 
  cursor_bookmark("b3") |> 
  body_add_par("Text for B3") |> 
  cursor_bookmark("c2") |> 
  body_add_par("Text for C2") |> 
  cursor_bookmark("d4") |> 
  body_add_par("Text for D4") |> 
  cursor_bookmark("betweenIpsum") |> 
  body_add_par("Text for between Lorem Ipsum")

print(doc, target = "out.docx")
```

There should be four pieces of text in the table and a piece between the two lorem ipsum paragraphs. But as you'll see, the text that should be in the table is placed after the table instead:
[out.docx](https://github.com/davidgohel/officer/files/13362546/out.docx)

- [x] Provide the results of R command `sessionInfo()`. It had to be executed after you loaded the packages used by your example. This will let me know what is your version of R and what are the versions of the packages you used in your example. 

```
> sessionInfo()
R version 4.2.1 (2022-06-23 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

Matrix products: default

locale:
[1] LC_COLLATE=German_Switzerland.utf8  LC_CTYPE=German_Switzerland.utf8   
[3] LC_MONETARY=German_Switzerland.utf8 LC_NUMERIC=C                       
[5] LC_TIME=German_Switzerland.utf8    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] officer_0.6.3

loaded via a namespace (and not attached):
 [1] compiler_4.2.1    R6_2.5.1          ragg_1.2.2        tools_4.2.1       rstudioapi_0.14  
 [6] uuid_1.1-0        xml2_1.3.3        zip_2.2.0         textshaping_0.3.6 systemfonts_1.0.4
[11] openssl_2.0.6     askpass_1.1      
```

- [x] Make sure you did checked you had the latest version of the package on CRAN (and on github if issue exists with CRAN version).

- [x] Make sure you searched in the *open* and *closed* issues on the github repository.

I believe #509 to be related, but as you point out, the syntax is incorrect as it doesnt assign results.

**Comment by davidgohel** [comment]
Hello,

This behavior is the expected one.

```
library(officer)

doc <- read_docx("~/Downloads/bm.docx")

doc <- cursor_bookmark(doc, "a1") |> 
  body_add_par("Text for A1", pos = "on") 

print(doc, target = "out.docx")
```

<img width="1085" alt="Capture d’écran 2024-01-20 à 12 29 07" src="https://github.com/davidgohel/officer/assets/4331618/41489fc2-d47b-4579-b823-9087786c7ad5">


`pos='on'` replace the black where is the bookmark by a new block. It does not replace nested block (as you expect, unfortunately). See https://ardata-fr.github.io/officeverse/officer-for-word.html#replacement



> We recommend using the package named ‘doconv’ instead. It allows you to manipulate Word to perform some difficult tasks such as updating calculated fields.
> 
> It allows to update the tables of contents but also to refresh if they exist references to the document properties. We can then use these references, they can be defined manually with Word or programmatically with officer::run_word_field() which must use the calculated field DOCPROPERTY.
> 
> This feature make possible to replace chunk of text and not only a whole paragraph.


In short, use `doconv::docx_update()` after `officer::set_doc_properties()` and use calculated fields in your template to provide the values that will be used in the fields of your template.

**Comment by aghaynes** [comment]
Thanks for the info, but to me it seems you have removed a very useful feature. This used to work, and it seems a strange decision to remove the functionality. I'm not referring to the table of contents, but a table within the body of the document, so the reference to `doconv::docx_update()` doesn't help here (although I can see potential for it elsewhere). 

To reiterate, it used to be that you could place a bookmark within a table and `officer` (version 0.4.3) would replace that text *within* the table, with the default `pos = "after"`. Now, the inserted text just appears after the table. I also don't really want to have to reprogram the whole table in R (e.g. with `flextable`, as there elements that, as far as I know, cannot be programmed - checkboxes).

**Comment by davidgohel** [comment]
> I'm not referring to the table of contents, but a table within the body of the document

Yes me too, I am not sure if you read the summary of the function. 

> Update **all fields** and table of contents of a Word document

You can also read example of `set_doc_properties()` that show how to set values 2 fields named "yoyo" and "glop".

My suggestion was to use **Word Fields** and update the values with `doconv::docx_update()`. 

1. set the fields in your template (Word fields, not bookmarks) 
2. define their values with `officer::set_doc_properties()` 
3. run `doconv::docx_update()`

"Something else" for https://github.com/CTU-Bern/SwissASR/issues/9

**Comment by aghaynes** [comment]
OK, I didn't get the connection. Thanks, I'll have to look into using fields instead...

**Comment by aghaynes** [comment]
Thanks for the hints, David. We've made the package functional again.

However, our CI (GHA) on mac is currently failing with the following message ([PR here](https://github.com/CTU-Bern/SwissASR/pull/10))
![image](https://github.com/davidgohel/officer/assets/6594028/f343d654-fdf8-4e58-8b44-4a47782c07d2)

Does it mean anything to you? I adapted our workflow to install libreoffice as per your [doconv workflow file](https://github.com/ardata-fr/doconv/blob/main/.github/workflows/R-CMD-check.yaml).

**Comment by davidgohel** [comment]
Hello

> Could it be the permissions you mention on the mac part of the doconv readme?

If you had installed 'Word', maybe... But `doconv:: docx_update()` is to be used locally, and won't work on Linux.

From `doconv:: docx_update()` manual: 

> Update all fields and table of contents of a Word document using "Microsoft Word". This function will not work if "Microsoft Word" is not available on your machine.

It can't work on CI/CD, I think Word and its license politic does not allow us to deploy on Github actions. 

I think there is no solution to your new problem.

**Comment by aghaynes** [comment]
OK, thanks.

Oddly the examples pass on windows (although I don't explicitly check that the word file exists or it's contents, so it may not work in the strictest sense, but R at least thinks it's worked)...

Thanks again :)

**Comment by davidgohel** [comment]
Oh really? OK, I'll have a look, I was expecting an error...

**Comment by aghaynes** [comment]
Yep. I had a quick look at your code... 
By the look of it, you silence the error on windows but not on mac...
https://github.com/ardata-fr/doconv/blob/e0bd5685013d7020eee407c0316661dc90b3673b/R/docx2pdf.R#L201
https://github.com/ardata-fr/doconv/blob/e0bd5685013d7020eee407c0316661dc90b3673b/R/docx2pdf.R#L227

![image](https://github.com/davidgohel/officer/assets/6594028/38803ac0-619d-46ca-bcb3-7fc13f863ef3)

I guess thats the cause...

**Comment by davidgohel** [comment]
Good catch thank you, this is not on purpose 😳
